File: android/src/main/java/duck/hansson/odd/android/app/ContactOwner.kt
@file:OptIn(ExperimentalMaterial3Api::class)

package duck.hansson.odd.android.app

import android.telephony.PhoneNumberUtils
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ListItem
import androidx.compose.material3.LocalMinimumInteractiveComponentEnforcement
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import duck.hansson.odd.shared.data.Owner
import duck.hansson.odd.shared.module.ContactActions
import java.util.Locale

@Composable
fun ContactOwner(
    actions: ContactActions,
    owner: Owner,
    onClose: () -> Unit,
    modifier: Modifier = Modifier
) {
    CompositionLocalProvider(LocalMinimumInteractiveComponentEnforcement provides false) {
        ModalBottomSheet(
            onDismissRequest = onClose,
            modifier = modifier
        ) {
            Text(
                text = owner.fullName,
                modifier = Modifier.padding(all = 16.dp),
                style = MaterialTheme.typography.titleLarge
            )

            owner.location?.let {
                ListItem(
                    modifier = Modifier.padding(bottom = 24.dp),
                    headlineContent = { Text(text = it.address) },
                    supportingContent = { Text("${it.zip} ${it.city}") }
                )
            }

            Column(
                modifier = Modifier
                    .padding(horizontal = 16.dp)
                    .padding(bottom = 16.dp),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                owner.phoneNumber?.let {
                    val formatted = PhoneNumberUtils.formatNumber(
                        it,
                        Locale.getDefault().country
                    )
                    ContactAction(
                        label = "Ring $formatted",
                        onAction = { actions.call(phoneNumber = it) }
                    )
                }

                owner.email?.let {
                    ContactAction(
                        label = "Meddela $it",
                        onAction = { actions.email(to = it) }
                    )
                }

                if (owner.noContacts) {
                    Text(
                        text = "Den här personen har inte lagt till några kontaktsätt.",
                        textAlign = TextAlign.Center
                    )
                }
            }
        }
    }
}

@Composable
fun ContactAction(
    label: String,
    onAction: () -> Unit,
    modifier: Modifier = Modifier
) {
    Button(
        onClick = onAction,
        modifier = modifier.fillMaxWidth(),
        colors = ButtonDefaults.buttonColors(containerColor = MaterialTheme.colorScheme.tertiaryContainer),
        border = CardDefaults.outlinedCardBorder(),
        contentPadding = PaddingValues(all = 16.dp)
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.onTertiaryContainer
        )
    }
}

File: ios/Odd/ContactOwner.swift
import KMMViewModelSwiftUI
import Shared
import SwiftUI

private struct InnerHeightPreferenceKey: PreferenceKey {
    static var defaultValue: CGFloat = .zero
    static func reduce(value: inout CGFloat, nextValue: () -> CGFloat) { value = nextValue() }
}

struct ContactOwner: View {
    let owner: Owner

    @EnvironmentViewModel var viewModel: AppViewModel
    @Environment(\.dismiss) private var dismiss

    @State private var sheetHeight: CGFloat = .zero

    var body: some View {
        VStack(alignment: .leading) {
            VStack(alignment: .leading) {
                Text(owner.fullName)
                    .font(.title)
                Text("Ägare")
            }
            .padding()

            if let location = owner.location {
                VStack(alignment: .leading) {
                    Text("\(location.address)")
                    Text("\(location.zip) \(location.city)")
                }
                .padding([.horizontal, .bottom])
            }

            if let phoneNumber = owner.phoneNumber {
                Button {
                    viewModel.contactActions.call(phoneNumber: phoneNumber)
                } label: {
                    Text("Ring \(phoneNumber)")
                        .frame(maxWidth: UIScreen.main.bounds.width)
                        .padding(.horizontal, 10)
                        .padding(.vertical, 5)
                        .font(.title3)
                }
                .padding(.horizontal)
                .buttonStyle(.bordered)
            }

            if let email = owner.email {
                Button {
                    viewModel.contactActions.email(to: email)
                } label: {
                    Text("Meddela \(email)")
                        .frame(maxWidth: UIScreen.main.bounds.width)
                        .padding(.horizontal, 10)
                        .padding(.vertical, 5)
                        .font(.title3)
                }
                .padding(.horizontal)
                .buttonStyle(.bordered)
            }

            if owner.noContacts {
                Text("Den här personen har inte lagt till några kontaktsätt.")
                    .frame(maxWidth: UIScreen.main.bounds.width)
                    .fixedSize(horizontal: false, vertical: true)
                    .multilineTextAlignment(.center)
            }
        }
        .padding(.vertical)
        .overlay {
            GeometryReader {
                Color.clear.preference(key: InnerHeightPreferenceKey.self, value: $0.size.height)
            }
        }
        .onPreferenceChange(InnerHeightPreferenceKey.self) { sheetHeight = $0 }
        .presentationDetents([.height(sheetHeight)])
        .presentationDragIndicator(.visible)
        .onChange(of: viewModel.state) { state in
            if !(state is AppState.Data) {
                dismiss()
            }
        }
    }
}
